File: Frontend/orbitavanya/server/routes/notifications.js
import { Router } from 'express';
import Notification from '../models/Notification.js';
import User from '../models/User.js';
import { requireAuth } from '../middleware/auth.js';
import { pushNotification } from '../utils/notify.js';

const router = Router();
router.use(requireAuth);

function toPublicNotification(n) {
  return {
    id: n._id,
    type: n.type,
    title: n.title,
    message: n.message,
    link: n.link,
    relatedId: n.relatedId,
    read: n.read,
    createdAt: n.createdAt,
  };
}

// ---------- GET /api/notifications — latest alerts for the signed-in user ----------
router.get('/', async (req, res) => {
  try {
    const notifications = await Notification.find({ user: req.userId })
      .sort({ createdAt: -1 })
      .limit(50);
    const unreadCount = await Notification.countDocuments({ user: req.userId, read: false });
    res.json({ notifications: notifications.map(toPublicNotification), unreadCount });
  } catch (err) {
    console.error('List notifications failed:', err.message);
    res.status(500).json({ error: 'Could not load notifications.' });
  }
});

// ---------- POST /api/notifications — create a custom alert ----------
// Lets any signed-in user raise their own reminder/alert, or notify a
// specific teammate about anything that isn't a meeting/task event.
router.post('/', async (req, res) => {
  try {
    const { title, message, link, userId } = req.body || {};
    if (!title) return res.status(400).json({ error: 'Title is required.' });

    let targetId = req.userId;
    if (userId) {
      const exists = await User.exists({ _id: userId });
      if (!exists) return res.status(400).json({ error: 'Target user not found.' });
      targetId = userId;
    }

    const notification = await pushNotification({
      userId: targetId,
      type: 'custom',
      title,
      message: message || '',
      link: link || '',
    });

    if (!notification) return res.status(500).json({ error: 'Could not create alert.' });
    res.status(201).json({ notification: toPublicNotification(notification) });
  } catch (err) {
    console.error('Create notification failed:', err.message);
    res.status(500).json({ error: 'Could not create alert.' });
  }
});

// ---------- PATCH /api/notifications/read-all ----------
router.patch('/read-all', async (req, res) => {
  try {
    await Notification.updateMany({ user: req.userId, read: false }, { $set: { read: true } });
    res.json({ ok: true });
  } catch (err) {
    console.error('Mark all read failed:', err.message);
    res.status(500).json({ error: 'Could not update alerts.' });
  }
});

// ---------- PATCH /api/notifications/:id/read ----------
router.patch('/:id/read', async (req, res) => {
  try {
    const notification = await Notification.findOneAndUpdate(
      { _id: req.params.id, user: req.userId },
      { $set: { read: true } },
      { new: true }
    );
    if (!notification) return res.status(404).json({ error: 'Alert not found.' });
    res.json({ notification: toPublicNotification(notification) });
  } catch (err) {
    console.error('Mark read failed:', err.message);
    res.status(500).json({ error: 'Could not update alert.' });
  }
});

// ---------- DELETE /api/notifications/:id ----------
router.delete('/:id', async (req, res) => {
  try {
    const result = await Notification.findOneAndDelete({ _id: req.params.id, user: req.userId });
    if (!result) return res.status(404).json({ error: 'Alert not found.' });
    res.json({ ok: true });
  } catch (err) {
    console.error('Delete notification failed:', err.message);
    res.status(500).json({ error: 'Could not delete alert.' });
  }
});

export default router;

File: app/routes/notifications.py
"""
api/routes/notifications.py
----------------------------
In-app notification endpoints — mirrors Node.js server/routes/notifications.js.

Endpoints:
  GET    /api/notifications           — list user's notifications (+ unreadCount)
  POST   /api/notifications           — create a custom alert
  PATCH  /api/notifications/read-all  — mark all as read
  PATCH  /api/notifications/:id/read  — mark one as read
  DELETE /api/notifications/:id       — delete one
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Optional

from bson import ObjectId
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.core.auth import get_current_user
from utils.db_client import get_collection

router = APIRouter(prefix="/notifications", tags=["notifications"])


def _to_public_notification(n: dict) -> dict:
    return {
        "id": str(n["_id"]),
        "type": n.get("type", ""),
        "title": n.get("title", ""),
        "message": n.get("message", ""),
        "link": n.get("link", ""),
        "relatedId": n.get("relatedId", ""),
        "read": n.get("read", False),
        "createdAt": n.get("createdAt", ""),
    }


@router.get("")
def list_notifications(current_user: dict = Depends(get_current_user)):
    notifs_col = get_collection("notifications")
    uid = current_user["_id"]
    notifications = list(
        notifs_col.find({"user": uid}).sort("createdAt", -1).limit(50)
    )
    unread_count = notifs_col.count_documents({"user": uid, "read": False})
    return {
        "notifications": [_to_public_notification(n) for n in notifications],
        "unreadCount": unread_count,
    }


class CreateNotifBody(BaseModel):
    title: str
    message: Optional[str] = ""
    link: Optional[str] = ""
    userId: Optional[str] = None


@router.post("", status_code=201)
def create_notification(
    body: CreateNotifBody,
    current_user: dict = Depends(get_current_user),
):
    if not body.title:
        raise HTTPException(400, "Title is required.")

    notifs_col = get_collection("notifications")
    users_col = get_collection("users")

    target_id = current_user["_id"]
    if body.userId:
        try:
            target_oid = ObjectId(body.userId)
        except Exception:
            raise HTTPException(400, "Invalid userId.")
        if not users_col.find_one({"_id": target_oid}):
            raise HTTPException(400, "Target user not found.")
        target_id = target_oid

    result = notifs_col.insert_one({
        "user": target_id,
        "type": "custom",
        "title": body.title,
        "message": body.message or "",
        "link": body.link or "",
        "relatedId": "",
        "read": False,
        "createdAt": datetime.now(tz=timezone.utc),
    })
    notification = notifs_col.find_one({"_id": result.inserted_id})
    if not notification:
        raise HTTPException(500, "Could not create alert.")
    return {"notification": _to_public_notification(notification)}


@router.patch("/read-all")
def mark_all_read(current_user: dict = Depends(get_current_user)):
    get_collection("notifications").update_many(
        {"user": current_user["_id"], "read": False},
        {"$set": {"read": True}},
    )
    return {"ok": True}


@router.patch("/{notif_id}/read")
def mark_read(
    notif_id: str,
    current_user: dict = Depends(get_current_user),
):
    try:
        oid = ObjectId(notif_id)
    except Exception:
        raise HTTPException(400, "Invalid notification ID.")

    notification = get_collection("notifications").find_one_and_update(
        {"_id": oid, "user": current_user["_id"]},
        {"$set": {"read": True}},
        return_document=True,
    )
    if not notification:
        raise HTTPException(404, "Alert not found.")
    return {"notification": _to_public_notification(notification)}


@router.delete("/{notif_id}")
def delete_notification(
    notif_id: str,
    current_user: dict = Depends(get_current_user),
):
    try:
        oid = ObjectId(notif_id)
    except Exception:
        raise HTTPException(400, "Invalid notification ID.")

    result = get_collection("notifications").find_one_and_delete(
        {"_id": oid, "user": current_user["_id"]}
    )
    if not result:
        raise HTTPException(404, "Alert not found.")
    return {"ok": True}
